### PR TITLE
Use libcxxabi and add tests for stdlib=libc++

### LINKED ISSUE #9: Test building C++ program
Would be good if a simple test C++ program were run in the build stage.

### PR DESCRIPTION
**Pull request by isuruf**
Fixes #9

### CODE REVIEW

**Comment by isuruf** [comment]
This is ready now

**Comment by jakirkham** [comment]
What's the benefit of building with libcxxabi?

**Comment by isuruf** [comment]
`libc++` is incomplete without an ABI library like `libc++abi`. If you try the example C++ sources, it will throw an error about undefined symbols when `-fno-exceptions` is not given. For C++ exceptions support this is needed.

**Comment by jakirkham** [comment]
I see. FWICT these tests only fail on Linux without these build changes. Are they intended to fail on macOS without these build changes as well?

File: recipe/test_sources/test_exceptions.cpp [hunk 1]
```diff
@@ -0,0 +1,11 @@
+#include <stdexcept>
+#include <iostream>
+
+int main() {
+    try {
+        throw std::runtime_error("hello");
+    } catch (std::runtime_error &e) {
+        std::cout << "hello" << std::endl;
```
Comment (jakirkham) [inline recipe/test_sources/test_exceptions.cpp]: Maybe it is worth changing `"hello"` to `e.what()` just to exercise the exception object a bit.

File: recipe/build.sh [hunk 2]
```diff
@@ -5,11 +5,42 @@ export CC=clang CXX=clang++
 cmake \
   -DCMAKE_INSTALL_PREFIX=$PREFIX \
   -DCMAKE_BUILD_TYPE=Release \
-  -DLLVM_ENABLE_RTTI=ON \
-  -DLLVM_INCLUDE_TESTS=OFF \
-  -DLLVM_INCLUDE_DOCS=OFF \
```
Comment (jakirkham) [inline recipe/build.sh]: These were probably holdovers from the LLVM package and I can't see why libcxx would use need them in the libcxx build. That said, I didn't see any warnings from CMake that I would have expected about these arguments being unused in the most recent [build]( https://travis-ci.org/conda-forge/libcxx-feedstock/builds/236719798 ) with this recipe. Though I'm not seeing them being used for anything in that build either. So they should be safe to drop I think. Would be good if others checked though.

File: recipe/build.sh [hunk 3]
```diff
@@ -5,11 +5,42 @@ export CC=clang CXX=clang++
...
+cmake \
+  -DCMAKE_INSTALL_PREFIX=$PREFIX \
+  -DCMAKE_BUILD_TYPE=Release \
+  -DLIBCXX_CXX_ABI=libcxxabi \
+  -DLIBCXX_CXX_ABI_INCLUDE_PATHS=../libcxxabi-${PKG_VERSION}.src/include \
+  ..
+
+make -j${CPU_COUNT}
+rm ${PREFIX}/lib/libc++.*
+make install
```
Comment (jakirkham) [inline recipe/build.sh]: Would `make install` automatically replace these? Admittedly it is good to be cautious here. Are there any other things we should be deleting in advance? Should we be doing this removal earlier?

File: recipe/compile_test.sh [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+    ./a.out
+done
+
+FILES=test_sources/*.cpp
+for f in $FILES
+do
+    clang++ -stdlib=libc++ -O2 -g $f
+    ./a.out
+    clang++ -stdlib=libc++ -std=c++11 -O2 -g $f
+    ./a.out
```
Comment (jakirkham) [inline recipe/compile_test.sh]: This is definitely a good idea. Though I wonder if we could have some tests that exercise the C++11 support. Maybe we could give them a different extension like `.cxx` for instance?

**Comment by jakirkham** [comment]
We still need a build `number` bump, right?

**Comment by jakirkham** [comment]
Added some more review comments. The changes seem like a good idea given the issues you have pointed out.

Would be nice to know the context this showed up in. For instance, did this occur when building Python packages with `libcxx` or something?

Will see if we can do something similar on Windows too in PR ( https://github.com/conda-forge/libcxx-feedstock/pull/11 ).

**Comment by jakirkham** [comment]
Also feel free to add yourself as a maintainer if you would like.

File: recipe/build.sh [hunk 5]
```diff
@@ -5,11 +5,42 @@ export CC=clang CXX=clang++
...
+cd libcxxabi-${PKG_VERSION}.src
+
+mkdir build
+cd build
+
+cmake \
+  -DCMAKE_INSTALL_PREFIX=$PREFIX \
+  -DCMAKE_BUILD_TYPE=Release \
+  -DLIBCXXABI_LIBCXX_PATH=../../ \
+  -DLIBCXXABI_LIBCXX_INCLUDES=../../include \
```
Comment (jakirkham) [inline recipe/build.sh]: It might be a little cleaner to rewrite these `../../` paths as `"$SRC_DIR"`.

File: recipe/build.sh (hunk 2 above)
Comment (jakirkham) [inline recipe/build.sh]: Actually it looks like these are proxies. So we should keep them, but we can use the `LIBCXX_` variables instead.

ref: https://github.com/llvm-mirror/libcxx/blob/6933299de0eb9d4528c39dc5462deb12a3c0ed50/CMakeLists.txt#L54
ref: https://github.com/llvm-mirror/libcxx/blob/6933299de0eb9d4528c39dc5462deb12a3c0ed50/CMakeLists.txt#L72

File: recipe/build.sh [hunk 6]
```diff
@@ -5,11 +5,42 @@ export CC=clang CXX=clang++
...
-  -DLLVM_INCLUDE_DOCS=OFF \
   ..
 
-make -j 8
+make -j${CPU_COUNT}
 make install
+cd ..
 
+# Build libcxxabi
+curl -L -O http://llvm.org/releases/${PKG_VERSION}/libcxxabi-${PKG_VERSION}.src.tar.xz
```
Comment (jakirkham) [inline recipe/build.sh]: We should add `curl` as a `build` dependency then.

File: recipe/build.sh (hunk 5 above)
Comment (jakirkham) [inline recipe/build.sh]: Also do `libcxx` and `libcxxabi` need to be nested inside each other like this or could they be placed next to each other.

Comment (isuruf) [inline recipe/build.sh]: Both need the internal headers of the other. So, I thought it's better to have them both in the same recipe.

File: recipe/build.sh (hunk 3 above)
Comment (isuruf) [inline recipe/build.sh]: For some reason, the libraries are not getting replaced in the second run. That's why they are manually replaced.

**Comment by isuruf** [comment]
> Would be nice to know the context this showed up in. For instance, did this occur when building Python packages with libcxx or something?

I tried the package with a simple hello world and clang++ didn't work.

File: recipe/build.sh (hunk 6 above)
Comment (isuruf) [inline recipe/build.sh]: Done

File: recipe/build.sh (hunk 2 above)
Comment (isuruf) [inline recipe/build.sh]: they are set from the LLVM config file, so the variables are safe to be removed.

File: recipe/build.sh (hunk 5 above)
Comment (jakirkham) [inline recipe/build.sh]: Right, I meant does the structure need to be such that the `libcxxapi` directory is inside of `libcxx` or can they directories be placed next to each other instead?

Comment (isuruf) [inline recipe/build.sh]: Ah. yes, they can be placed next to each other. Just need the paths set as in above.

File: recipe/build.sh (hunk 2 above)
Comment (jakirkham) [inline recipe/build.sh]: Though don't we still want to exclude the docs somehow? How are they getting excluded if we aren't setting these?

**Comment by jakirkham** [comment]
I tried to incorporate similar changes into the Windows build. Though ran into a configuration issue where CMake was looking in the wrong place for CMake modules. Tried to fix it, but it seems to be sort of broken still. Have PR'd the changes against your branch in case you have a better idea of how to fix this. ( https://github.com/isuruf/libcxx-feedstock/pull/1 )

**Comment by isuruf** [comment]
@jakirkham, can we have this merged and work on a new PR for windows here ? (I've sent a PR here, https://github.com/jakirkham/libcxx-feedstock/pull/1)

**Comment by jakirkham** [comment]
Yep, LGTM. Thanks @isuruf.

**Comment by jakirkham** [comment]
Here is the new Windows PR ( https://github.com/conda-forge/libcxx-feedstock/pull/13 ).

**Pull request merged**